File: 740-delete-and-earn/delete-and-earn.java
class Solution {
    int[] dp;
    public int deleteAndEarn(int[] nums) {
      Map<Integer,Integer> freq = new HashMap<>();
      dp = new int[nums.length];

      Arrays.fill(dp , -1);

      for(int num : nums){
        freq.put(num , freq.getOrDefault(num , 0) + 1);
      }

      List<Integer> arr = new ArrayList<>(freq.keySet());
      Collections.sort(arr);

      return solve(0 , arr , freq);
    }

    private int solve(int i , List<Integer> arr , Map<Integer,Integer> map){
      if(i >= arr.size()){
        return 0;
      }

      if(dp[i] != -1) return dp[i];

      int np = solve(i+1 , arr , map);
      int curr = arr.get(i);
      int p = curr * map.get(curr);

      int next = i + 1;

      if(next < arr.size() && arr.get(next) - 1 == curr ){
        p += solve(i+2 , arr , map);
      } else{
        p += solve(i+1 , arr , map);
      }

      return dp[i] = Math.max(np , p);
    }

}

File: 740-delete-and-earn/delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        m = {}

        for num in nums:
            m[num] = 1 if num not in m else m[num] + 1
        
        nums = list(m.keys())
        nums.sort()
        n = len(nums)

        dp = [0 for _ in range(n+2)]

        for i in range(n-1, -1, -1):
            np = dp[i+1]
            curr = nums[i]
            p = curr* m[curr]

            if i+1 < n and nums[i+1] == curr + 1:
                p += dp[i+2]
            else:
                p += dp[i+1]
            
            dp[i] = max(p, np)
        
        return dp[0]

                

        